File: inst/js/widgetTimeDB.js
$(document).ready(function() {

  function initiateWidgetTimeDB(widget) {

    widget = $(widget);

    widget.find('.tbEdit').timeEntry({show24Hours: true});

    if (widget.find('.divNoEdit').data('onlydataentry') == 1) {
      widget.find('.divEdit').show();
      widget.find('.divNoEdit').hide();
    } else {
      widget.find('.divEdit').hide();
    }


    if (widget.find('.divNoEdit').data('readonly') == 1) {
      widget.find('.divNoEdit').css('background-color', '#E7E7E7');
    }

    if (widget.find('.divNoEdit').data('isreqd') == 1) {
      if (widget.find('.divNoEdit p').text() === '') {
        widget.find('.divVal').append('<p>Text cannot be missing</p>');
      }
    }

  }

  // var inputTxt = prompt('Enter text: ', '');
  //$('.divNoEdit p').text(inputTxt);

  //var initVal = $(document).find('.divNoEdit p').text();
  //var finalVal = '';

  $(document).on('dblclick', '.widgetTimeDB .divNoEdit', function(evt) {
    var mother = $(this).parents('.widgetTimeDB');
      if (mother.find('.divNoEdit').data('readonly') == 0) {
        mother.find('.divNoEdit').hide(150);
        mother.find('.divEdit').show(150);
      }
      mother.find('.tbEdit').val(mother.find('.divNoEdit p').text());
      if (mother.find('.divNoEdit').data('readonly') == 1) {
        mother.find('.btnUpdate').trigger('click');
      }    //initVal = mother.find('.divNoEdit p').text();

  }); // end of double click

  $(document).on('click', '.widgetTimeDB .btnUpdate', function(evt) {

    var mother = $(this).parents('.widgetTimeDB');
    //var el = $(evt.target).parents('.widget');
    var finalVal = mother.find('.tbEdit').val();

    mother.find('.divVal').empty();

    if (mother.find('.divNoEdit').data('isreqd') == 1) {
      if (finalVal === '') {
        mother.find('.divVal').append('<p>Text cannot be missing.</p>');
        return null;
      } else {
        mother.find('.divVal').empty();
      }
    }

    mother.find('.divNoEdit p').text(finalVal);
    mother.find('.divNoEdit').css('border-color', 'red');
    mother.find('.divEdit').hide(150);
    mother.find('.divNoEdit').show(150);
    mother.trigger('change');

  }); // end of click on update button

  $(document).on('click', '.widgetTimeDB .btnCancel', function(){

    var mother = $(this).parents('.widgetTimeDB');
    mother.find('.tbEdit').text('');
    mother.find('.divEdit').hide(150);
    mother.find('.divNoEdit').show(150);

    if (mother.find('.divVal').html() !== '') {
      mother.find('.divVal').empty();
    }

  }); // end of click on cancel button

  // data from server to client on updating server side value of widget or on updating changes to database (receives server side updateWidget function output)

  Shiny.addCustomMessageHandler('sendToWidgetTimeDB', function(data) {

    if (data.isUpdatedInDB[0] === true) {
      $('#' + data.id[0]).find('.divNoEdit').css('border-color', 'gray');
    }

    if (data.value[0] !== '') {
      $('#' + data.id[0]).find('.divNoEdit p').text(data.value[0]);
    }

    if (data.msg[0] !== '') {
      $('#' + data.id[0]).find('.divVal').append('<p>' + data.msg[0] + '</p>');
    }

  });

  Shiny.addCustomMessageHandler('validateWidgetTimeDB', function(data) {

    var mother = $('#' + data.id[0]).parents('.widgetTimeDB');
    mother.find('.tbEdit').val('');
    $('#' + data.id[0]).find('.divVal').append('<p>' + data.message[0] + '</p>');

  });

  var widgetTimeDBBinding = new Shiny.InputBinding();

  $.extend(widgetTimeDBBinding, {

    find: function(scope) {
      return $(scope).find('.widgetTimeDB');
    },

    initialize: function(el) {
      initiateWidgetTimeDB(el);
    },

    getValue: function(el) {
      if ($(el).find('.divNoEdit p').text() !== '') {
        return  $(el).find('.divNoEdit p').text();
      } else {
        return null;
      }
    },

    getType: function() {
      return 'shinyDBWidget.widgetTimeDBBinding';
    },

    subscribe: function(el, callback) {
      $(el).on('change.widgetTimeDBBinding', function(e) {
        callback();
      });
    },

    unsubscribe: function(el) {
      $(el).off('.widgetTimeDBBinding');
    }
  });

  Shiny.inputBindings.register(widgetTimeDBBinding, 'shinyDBWidget.widgetTimeDBBinding');

}); //end of ready

File: R/widgetTimeDB.R
#' @export
widgetTimeDB <- function(inputId, label, value = NULL, isRequired = FALSE, readOnly = FALSE, onlyDataEntry = FALSE, defaultVal = NULL) {

  isRequired <- as.integer(isRequired)
  readOnly <- as.integer(readOnly)
  dataEntry <- as.integer(onlyDataEntry)

  if (!is.null(defaultVal) && (is.null(value)))
    value <- as.character(defaultVal)

  if (!is.null(value) && !isTime(value)) stop(paste0("id = ", inputId, ": Please enter correct time."))

  if (dataEntry == 1) {
    readOnly <- 0
    value <- ""
  }

  shiny::addResourcePath("js", system.file("js", package = "shinyDBWidget"))
  shiny::addResourcePath("css", system.file("css", package = "shinyDBWidget"))
  shiny::addResourcePath("timeentry", system.file("shared/timeentry", package = "shinyDBWidget"))

  tagList(singleton(tags$head(
    tags$link(rel="stylesheet", href="css/widgetDB.css", type="text/css", media="all"),
    tags$link(rel="stylesheet", href="timeentry/jquery.timeentry.css", type="text/css", media="all"),
    tags$script(src="timeentry/jquery.plugin.min.js"),
    tags$script(src="timeentry/jquery.timeentry.min.js"),
    tags$script(src="js/widgetTimeDB.js")
  )),
  tags$body(tags$div(id = inputId, class = "widgetTimeDB widgetDB",
                     tags$p(tags$b(label)),
                     tags$div(class = "divNoEdit", `data-isReqd` = isRequired, `data-readOnly` = readOnly, `data-onlyDataEntry` = dataEntry, p(value)),
                     tags$div(class = "divEdit",
                              tags$input(class = "tbEdit", type = "text"),
                              tags$span(title = "Decrement", class= "timeEntry-control", style = 'display: inline-block; background: transparent url(spinnerDefault.png) no-repeat scroll 0px 0px; width: 20px; height: 20px;'),
                     tags$div(class = "divBtn",
                              tags$input(type = "button", class = "btn btnUpdate", value = "Update"),
                              tags$input(type = "button", class = "btnCancel btn", value = "Cancel"))),
                     tags$div(class = "divVal"))))
}

#' @export
updateWidgetTimeDB <- function(session, inputId, isUpdatedInDB = FALSE, value = NULL) {

  # This function will take either the isUpdatedInDB or value or both of them and change the widget accordingly:
  # If isUpdatedInDB is true then it will revert back the color of border of the .divNoEdit p to gray. If false no change will be made. isUpdatedInDB will come from the server.R at the time when the edited values have entered into the DB successfully.
  # If value is not null, then it will be used to change the .divNoEdit p value into the new value, also the value of .tbEdit is changed into "".

  # This function will be called from server and will make changes to the widget denoted by the inputId accordingly.
  # It will make use of the session$sendCustomMessage and Shiny.addCustomMessageHandler

  # session$sendCustomMessage(type = "name", msg)
  #browser()

  if (!is.null(value) && !isTime(value)) {
    msg <- "Time required on update"
    value <- ""
  } else msg <- ""

  session$sendCustomMessage(type = "sendToWidgetTimeDB", jsonlite::toJSON(list(id = inputId, isUpdatedInDB = isUpdatedInDB, value = value, msg = msg)))

}

#' @export
validateWidgetTimeDB <- function(session, input, inputId, condition, message) {

  # The problem with the shiny::validate is that it has to be placed inside a reactive object and that it will output the message inside the output component.

  # There will be a tremendous effort required if I want to add validation along with the constructor function of the widget, as the validation code will have to be translated to javascript before running it, it will need a new DSL, which is beyond me. I will put the validation code in server side and will pass the meaasge using session$sendCustomMessage and will handle in Shiny.addCustomMessageHandler. The problem will be to make this function run as soon as the input$inputId changes and to make the input$inputId value chage to NULL if validation is found to be wrong.

  # condition: it will be denoted as a valid R expression which will output TRUE/FALSE, if TRUE it will be validated else it will be invalidated, with message passed to the javascript routine and printed in divVal div of the widget and input$inputId changed to NULL.
  #browser()

  if (!isTime(input[[inputId]])) return(NULL)

  condition <- substitute(condition)

  isValidated <- eval(condition, envir = list(`.` = input[[inputId]]), enclos = parent.frame())

  if (isValidated == FALSE) {
    session$sendCustomMessage(type = "validateWidgetTimeDB", jsonlite::toJSON(list(id = inputId, message = message)))
    #input[[inputId]] <- NULL
  }

  return(NULL)
}

#' @export
registerWidgetTimeDB <- function(fun = function(data, ...) return(data)) {
  environment(fun) <- parent.frame()
  shiny::registerInputHandler(type = "shinyDBWidget.widgetTimeDBBinding", fun, force = TRUE)
}
